modules/c++/math.linear/include/math/linear/MatrixMxN.h

```diff
@@ -39,7 +39,17 @@ template<typename _T> bool equals(const _T& e1, const _T& e2)
 
 template<typename _T> inline bool equals(const _T& e1, const _T& e2, _T eps)
 {
+#ifdef _MSC_VER // Visual Studio
```

Comment:
Given that the compiler knows that it's a constant at compile time, it's almost certainly going to optimize it out when compiling for release, so there's not going to be a performance hit.  But I do agree that it's unsatisfying the way it's structured now.

At first I liked the `condition` class on Stack Overflow, but thinking about it a bit more, aside from the advantage of skirting around the compiler warning, it's still the same thing... you have an `if` statement with a condition that the compiler can evaluate at compile time based on which template specialization it picks and then presumably inlines it away.

I've done some stuff with boost back in the day where they had a `true_type` and `false_type`.  Then you made a function overloading where the last parameter was either this true or false type, they had a similar condition type object which presumably again used template specialization to return the correct type... but then here there truly wasn't an `if` statement; it was the compiler choosing the correct function overloading which of course happened at compile time.  Something notionally like:

```
template <typename T>
bool equals(const T& e1, const T& e2, T eps, boost::false_type)
{
    return std::abs(e1 - e2) < eps;
}

template <typename T>
bool equals(const T& e1, const T& e2, T eps, boost::true_type)
{
    if (e1 == std::numeric_limits<T>::infinity() ||
        e1 == -std::numeric_limits<T>::infinity())
    {
        return e1 == e2;
    }

    return equals(e1, e2, eps, boost::false_type);
}

template <typename T>
bool equals(const T& e1, const T& e2, T eps)
{
    return equals(e1, e2, eps, boost::conditional(std::numeric_limits<T>::has_infinity));
}
```

We're (unfortunately) not heavily using boost right now, though I'm open to changing that when we have a compelling reason... it's mostly due to momentum that we hadn't used it in the past for these projects.  But here, I think we can pretty easily accomplish what we want without boost by still using template specialization.  I didn't actually compile this but the syntax should be close:

```
template <bool b>
struct Equals
{
    template <typename T>
    static bool compare()(const T& e1, const T& e2, T eps)
    {
        return std::abs(e1 - e2) < eps;
    }
};

template <>
struct Equals<true>
{
    template <typename T>
    static bool compare()(const T& e1, const T& e2, T eps)
    {
        if (e1 == std::numeric_limits<T>::infinity() ||
            e1 == -std::numeric_limits<T>::infinity())
        {
            return e1 == e2;
        }
        
        return Equals<false>::compare(e1, e2, eps);
    }
};

template <typename T>
bool equals(const T& e1, const T& e2, T eps)
{
    return Equals<std::numeric_limits<T>::has_infinity>::compare(e1, e2, eps);
}
```

Obviously someone needs to understand more C++ to see what's going on here, and it's increased the number of lines of code, but I think this is legible if you want to put something like this together.  Make sure to add unit tests if there aren't some already! :o)